Move IImageMonikerService to EditorFeatures.Wpf

davidwengier opened the pull request:
As per [this comment](https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/Core/Shared/Extensions/GlyphExtensions.cs#L26-L28) we want to avoid ImageMoniker references in VS for Mac but unfortunately `IImageMonikerService` did not live up to this goal. This PR obsoletes it and moves it to EditorFeatures.Wpf.

Ideally I would like to completely remove it from EditorFeatures completely. I searched //index and gtihub and there doesn't seem to be any implementations of it, but I assume we need to obsolete it first. **Please let me know if that's not true, and I can just delete it.**

/cc @KirillOsenkov

davidwengier commented on src/EditorFeatures/Core.Cocoa/Microsoft.CodeAnalysis.EditorFeatures.Cocoa.csproj:
```diff
@@ -17,6 +17,7 @@
     <Compile Include="..\Core.Wpf\SignatureHelp\**\*.cs" LinkBase="SignatureHelp" />
     <Compile Include="..\Core.Wpf\Suggestions\**\*.cs" LinkBase="Suggestions" />
     <Compile Include="..\Core.Wpf\IDebuggerTextView2.cs" Link="IDebuggerTextView2.cs" />
+    <Compile Include="..\Core.Wpf\Tags\IImageMonikerService.cs" LinkBase="Tags" />
```
I don't love the idea of adding a single interface to Cocoa.dll which will never be implemented, but this avoids peppering `#if !COCOA` around the place so its a bit neater. Happy to change to that model if people prefer though.

KirillOsenkov replied:
I think this is fine

jasonmalinowski replied:
So why is this being added like this? I would have thought the whole point of this was to get rid of this on the Cocoa side of things.

davidwengier replied:
This is just to avoid putting various `#if COCOA` blocks in here: https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSourceProvider.cs#L64

I'm happy to do that if its preferred, I don't feel strongly either way. This still means that someone has to try extra hard to use it on the cocoa side, rather than accidentally doing it because they reference EditorFeatures, but certainly the `#if`s are a stronger statement.

KirillOsenkov replied:
Oh, I hadn't realized that this way we'll still have ImageMoniker in EF.Cocoa.dll. Maybe it's indeed better to do `#if COCOA` on the Wpf side? If we can remove the Imaging reference entirely from Cocoa.dll that would be an extra safeguard to avoid accidentally using it.

KirillOsenkov replied:
Hmm sure would be nice to get rid of the interop reference here:

![image](https://user-images.githubusercontent.com/679326/114459260-2352e680-9b95-11eb-8e98-0827685af063.png)

jasonmalinowski replied:
So it's not just EditorFeatures.Cocoa having the reference, the editor APIs themselves still do: even on the cocoa side, ISuggestedAction.ImageMoniker exists. The EditorFeatures layer _cannot_ get rid of ImageMoniker references on it's own -- we need the Editor to fix layering first.

davidwengier replied:
Yeah, sadly we can't entirely get rid of ImageMoniker, but at least with this change we don't have it as an embedded type in EditorFeatures.dll itself.

jasonmalinowski replied:
As long as @KirillOsenkov is OK with this.

KirillOsenkov replied:
Ah, yes, ISuggestedAction. I knew we can't have nice things! And here I was hopeful for a brief moment. OK yeah if we can't get rid of the reference anyway than I'm OK with whatever, sorry for the confusion.

KirillOsenkov replied:
Will we have the embedded type in EF.Cocoa.dll?

davidwengier replied:
No, that was already turned off in https://github.com/dotnet/roslyn/pull/51872

davidwengier commented on src/EditorFeatures/Core.Wpf/Tags/DefaultImageMonikerService.cs:
```diff
@@ -12,7 +12,7 @@
 using Microsoft.VisualStudio.Imaging.Interop;
 
-namespace Microsoft.CodeAnalysis.Editor.Tags
+namespace Microsoft.CodeAnalysis.Editor.Tags.Implementation
```
I would prefer to have this be `Microsoft.CodeAnalysis.Editor.Tags.Wpf` to be clear about where it lives, but it felt weird to do that and then share it with cocoa.dll. If we change to using `#if !COCOA` then we're free-er to change this.

sharwell reviewed (changes requested):
Question before merge:

❓ What happens if we leave the type in its current location, and just don't use this service in VS Mac?

Therzok commented:
Would need to ensure the type is not JITted at all, because Mono does not support embedded interop types.

KirillOsenkov commented:
Note the VSMac constraint isn't the only reason this would be desirable. Imaging and ImageMoniker are VS implementation details that unfortunately have leaked into the shared API layer (EditorFeatures). Ideally they'd only belong in the VS layer (MS.VS.LanguageServices), and not impose this type on all other hosts of Roslyn. 

So potentially with this PR we could completely remove the Imaging reference from EditorFeatures.dll, thus cleaning up the dependencies and making sure VS implementation details aren't leaking into shared.

sharwell commented:
@KirillOsenkov if this type didn't already ship, it would be a different discussion. Since the type already shipped, the question changed from "where should this type belong" to "can we make the original scenario work without moving the type"? It sounds like the answer is yes.

Therzok commented:
TypeForwardedFrom should help with moving this type

davidwengier commented:
I see this as another step in preventing Roslyn from breaking VS Mac inadvertently.

Yes we can "just don't use this service" but it's fragile. Moving it to a better place per Kirill's comment means it's harder for a dev to come along and add a usage without knowing the context. It means a developer dealing with a build failure in Roslyn rather than in a VS Mac insertion at some unknown future point.

davidwengier commented:
Would love to delete the old interface entirely to make it even more explicit a move:
https://grep.app/search?q=IImageMonikerService
http://index/?query=%60IImageMonikerService

sharwell commented:
I don't think this one is worth the compat break, as it's easy to fix an issue if it ever arose.

KirillOsenkov commented:
My opinion is that it's worth the break because it will allow a cleaner layering and remove a VS implementation detail from shared API.

Roslyn is breaking consumers all the time, e.g. updating to a newer Editor, Threading or other dependency requires consumers to recompile. And the only currently known consumer that will be impacted is begging you to do it (VSMac). I'm sure OmniSharp, LSP or others will prefer to have a cleaner layering then preserve an obscure unused API. We're not renaming ISemanticModel here.

It's a bit unfortunate and ironic that Roslyn has so much API surface that is internal, and of all things this one had to be a public API. 

I'm curious what are @CyrusNajmabadi and @jasonmalinowski thoughts on this.

jasonmalinowski commented:
Alright so a few thoughts here:

I agree ImageMoniker isn't something that is great to have in cross-plat code. That said, moving this type can potentially break third-party extenders which I don't think we want to do in 16.10, when we have an opportunity coming up to make breaking changes. And especially breaking changes around interop assemblies and embedding policies. If we just wanted to move the API unilaterally in 17.0 that's probably OK with me, since anybody using this type _also_ needs to make other changes for 17.0.

I'd otherwise say if we need to make a change for 16.10, let's just stop embedding the ImageMoniker assembly, which we can easily do. And that might be better anyways: as it stands our EditorFeatures.Cocoa project builds against Microsoft.VisualStudio.Language.Intellisense which has ImageMoniker already in it, so I don't see much value in moving this if we have other places we are already shipping ImageMoniker using code on Mac that are far more likely to get ran and loaded. Do we need to fix that too? Absolutely, but let's not pretend this is the only place we need to fix things.

KirillOsenkov commented:
I'm perfectly fine with making the API change for 17.0.

I think @davidwengier tried to stop embedding, but it spread virally, he can tell in more detail.

davidwengier commented:
> I think @davidwengier David Wengier FTE tried to stop embedding, but it spread virally, he can tell in more detail.

I spoke to Jason offline and he has pointed me to another potential solution to this, so will try again.

I think we should do both though: Remove the API in 17, and disable embedded interop more broadly if we can.

sharwell commented:
> I think we should do both though: Remove the API in 17, and disable embedded interop more broadly if we can.

For 17, interop embedding is generally disabled. One exception is `ImageMoniker`, which will still be embedded.

davidwengier commented:
Right, but we can still not embed it in Roslyn.

jasonmalinowski commented:
And if we move this in 17 no matter what any sort of disabling policy then can truly apply to EditorFeatures.Cocoa only,

sharwell commented:
> Right, but we can still not embed it in Roslyn.

It is my understanding that the assembly defining this structure is not available at runtime, so failure to embed will directly translate to an assembly load error.

davidwengier commented:
Rebased and targeting release/dev17.0

I also restored the original namespace and removed a couple of now-redundant references.

> It is my understanding that the assembly defining this structure is not available at runtime, so failure to embed will directly translate to an assembly load error.

Embedding or not is moot now, but maybe that statement is true on Windows, I don't know, but VS Mac already ships with ImageMoniker not embedded and the assemblies load fine ¯\\\_(ツ)_/¯

CyrusNajmabadi commented on src/EditorFeatures/Core/PublicAPI.Shipped.txt:
```diff
@@ -1,8 +1,2 @@
 Microsoft.CodeAnalysis.Editor.Peek.IPeekableItemFactory
 Microsoft.CodeAnalysis.Editor.Peek.IPeekableItemFactory.GetPeekableItemsAsync(Microsoft.CodeAnalysis.ISymbol symbol, Microsoft.CodeAnalysis.Project project, Microsoft.VisualStudio.Language.Intellisense.IPeekResultFactory peekResultFactory, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Collections.Generic.IEnumerable<Microsoft.VisualStudio.Language.Int ...
-Microsoft.CodeAnalysis.Editor.Tags.ExportImageMonikerServiceAttribute
```
do we need to discuss this breaking change with anyone?  or is the presumption that it was never ok for anyone to use this but us?

KirillOsenkov replied:
we just don’t know of anyone who depends on it, so we don’t know who to discuss this with

CyrusNajmabadi replied:
Do F# or TS use it?

KirillOsenkov replied:
nope, we checked http://index and grep.app

jasonmalinowski replied:
And yes, this was _OK_ to use as a public API, it's just that as best we can tell nobody ever needed it in the end.

CyrusNajmabadi replied:
wfm

jasonmalinowski commented on src/EditorFeatures/Core/Microsoft.CodeAnalysis.EditorFeatures.csproj:
```diff
@@ -30,9 +30,7 @@
     <PackageReference Include="Microsoft.VisualStudio.Text.UI" Version="$(MicrosoftVisualStudioTextUIVersion)" />
     <PackageReference Include="Microsoft.VisualStudio.Threading" Version="$(MicrosoftVisualStudioThreadingVersion)" />
-    <PackageReference Include="Microsoft.VisualStudio.SDK.EmbedInteropTypes" Version="$(MicrosoftVisualStudioSDKEmbedInteropTypesVersion)" />
     <PackageReference Include="Microsoft.VisualStudio.ImageCatalog" Version="$(MicrosoftVisualStudioImageCatalogVersion)" />
```
Can we get rid of ImageCatalog as well?

davidwengier commented:
Alternatively we could keep the service around and still remove the imaging stuff from EditorFeatures: https://github.com/dotnet/roslyn/pull/52595
